server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java

```diff
@@ -265,6 +266,24 @@ public CopyTo copyTo() {
         return copyTo;
     }
 
+    private String readComplexJsonElement(XContentParser parser) throws IOException {
+        int tokenDepth = 10;
+        XContentParser.Token closingToken = null;
+        if (parser.currentToken() == XContentParser.Token.START_ARRAY) {
+            closingToken = XContentParser.Token.END_ARRAY;
+        } else if (parser.currentToken() == XContentParser.Token.START_OBJECT) {
+            closingToken = XContentParser.Token.END_OBJECT;
+        } else return null;
+
+        StringBuilder builder = new StringBuilder();
+        for (int i = 0; i < tokenDepth
+            && parser.currentToken() != null
+            && parser.currentToken() != closingToken; i++) {
+            builder.append(parser.text());
```

Comment:
we may have problem here. `parser.text()` does not behave nicely in case of start/end of objects/arrays
https://github.com/elastic/elasticsearch/blob/master/libs/x-content/src/main/java/org/elasticsearch/common/xcontent/json/JsonXContentParser.java#L81-L86

I did not find how to read the string value for delimiters in a content type in this api https://github.com/elastic/elasticsearch/blob/92b2e1a2091ce30d6267029b7b7b19d4b6690d64/libs/x-content/src/main/java/org/elasticsearch/common/xcontent/XContentParser.java 
We can hardcode the json values for those delimiters, but that would not work with other types of contents, which may be a problem.

A dirty alternative would be using XContentParser.skipChildren() to get an offset in a parser buffer of beginning and ending of the serialized object/array, but we will have to handle the case of buffer refresh in case objects/array is big enough, in which case we cannot just take interval from one index in buffer to another. We can try to go around it by reading until certain max number of characters, but then we also risk hitting the right edge of buffer when currently cached buffer ends before the last char of the object/array. 

Not really sure what should we prioritize here - support of multiple XContents or simple and reliable parsing in case of json

Reply:
@sandmannn  What about instead all this code just use `AbstractXContentParser::readValue` method?  And if the value is NOT `null` and we don't get an `error` we can throw `MapperParsingException` with a value; otherwise `MapperParsingException` with just doc ID. 

Yes, there is a potential risk that a value is too long, but I think we can live with it.

Reply:
About the tests, the user that originally had created this issue, had a problem with a value of `{}` for a keyword type field.  It is good to have a test for this case.

Reply:
@mayya-sharipova  `AbstractXContentParser::readValue` would be a nice solution, unfortunately looks like it is has a package level access modifier, and thus not accessible from outside `org.elasticsearch.common.xcontent.support` package. We can try to either copy code of `readMap`, `readList` and `readValue` out or make those methods public, yet I am not sure why they were structured with this access level in the first place and what disadvantage such extended access may have.

Reply:
@sandmannn  Let's for now make the method `AbstractXContentParser::readValue` public, and see that it works well with tests. After that we can ask somebody from elasticsearch Core/Infra  team to review the PR, and if they are ok with this change, we can keep it.

Reply:
@mayya-sharipova  before adjusting the tests, I would like to explicitly align on one point: as described in this assert https://github.com/elastic/elasticsearch/pull/42321/files#diff-e10f66b49e08431ab42044d98647dadbR174 the resulting preview of the value does not have a json structure anymore. It provides some information about the actual parsed structure, but it is not showing its exact character structure, e.g. `{inner_field=value} instead of {"inner_field":"value"}` and potentially more weird stuff for nested structures.

I don't really see a better alternative, but lets align that we are fine with this before finalizing the PR.

Reply:
@sandmannn  I think this is fine, we are just making best effort to show the value.